### PR TITLE
PUPIL-249 feat(oak-layout): add oak layout to quiz

### PR DESCRIPTION
**Pull request by benprotheroe**
Add quiz layout to quiz view

## How to test

1. Go to https://deploy-preview-2207--oak-web-application.netlify.thenational.academy/pupils/programmes/english-secondary-ks4-aqa/units/macbeth-and-masculinity-the-struggle-for-power/lessons/duncan-as-a-father-figure
2. Click on quizzes
3. You should see quiz layout

## Screenshots

How it used to look (delete if n/a):
{screenshots}

How it should now look:
{screenshots}

## Checklist

- [x]  Use oakLessonLayout
- [x]  MCQs display ‘select x answers’
- [x]  Question count
- [x]  Back button and header
- [x]  Feedback in bottom nav
- [x]  Different colour backgrounds based on starter or exit quiz
- [x]  mobile, tablet browser views
- [x]  final button is ‘continue lesson’ instead of ‘next question’

### CODE REVIEW

File: src/components/PupilComponents/QuizRenderer/QuizRenderer.tsx
Comment (kimon-satan) [inline src/components/PupilComponents/QuizRenderer/QuizRenderer.tsx:75]: It troubles me that the encapsulation is now broken for this component and  PupilQuizView. Can you turn this id into a prop and have pupilQuizView pass it ?

File: src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx
Comment (kimon-satan) [inline src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx:37]: I feel this is pushing things in terms of readability . Could you pop this in its own component (can be inside this file)

File: src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx
Comment (kimon-satan) [inline src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx:41]: same issue with generic id

File: src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx
Comment (kimon-satan) [inline src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx:58]: try to avoid type assertion as it really is no protection at all. In this case you could write a simple conditional or write a custom type guard for CurrentSection (which should be renamed)

File: src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx
Comment (kimon-satan) [inline src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx:22]: lets call this something more specific like CurrentQuiz

File: src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx
Comment (kimon-satan) [inline src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx:80]: Redundant use of flex here

**Review by kimon-satan — CHANGES_REQUESTED** [review]
A few small things to pick up .

**Comment by messmike** [comment]
Design feedback:
- The bottom nav bar isn't currently fixed to the bottom of the browser window. Background & container should always take up the height of the window, with the nav bar fixed to the bottom. May need additional space applied so it always appears that way. E.g:
Screenshot 2024-01-23 at 1 21 36 pm
- The height of the bottom nav looks a little too large for the desktop, tablet & mobile breakpoints when there is no feedback present. Checked the designs, and the height should be 80px (currently 100px).
- The button in the bottom nav should be centre-aligned vertically across all breakpoints.
- For mobile, the button in the bottom nav should be aligned on the right, and the gleap button should be positioned to not obstruct anything in the bottom nav if that is possible to do? Here I have added 60px bottom margin to the gleap: 
Screenshot 2024-01-23 at 2 12 40 pm
- The question stem heading colour token looks incorrect, should use `text-primary`
- For MCQ questions in the starter and exit quizzes, the question stem and radio buttons middle area on mobile (max width) look to have more space on the left and right (margins) when compared to the designs. In the designs, the margins for mobile are set to 16px.
- The middle area (question stem and radio buttons/checkboxes/text input etc) on desktop across all questions in the starter and exit quiz sections (max width) should be set to 8cols for desktop (848px in the designs), currently at 738.359px. E.g: 
Screenshot 2024-01-23 at 2 24 04 pm There looks to be different widths across the quiz questions for both sections applied currently. 
- For the 2nd question of the starter quiz, and the 1st question of the exit quiz the space between the question stem heading and the 'Type your answer here' text should be using 'space-between-xl' for desktop, 'space-between-l' for tablet, and 'space-between-m' for mobile (same should be applied across the quizzes) 
- For the 2nd question of the starter quiz, and the 1st question of the exit quiz there is some shift in the size of the text input when submitting the question on desktop, and when the icon appears when in a valid or invalid state. The text input width shouldn't increase in width ideally across states. I think the width should be increased to something like 500px as a default for desktop. The text input should span the full width for tablet and mobile. E.g: 
Screenshot 2024-01-23 at 1 55 44 pm
- For the 3rd & 4th questions of the exit quiz, there is some shift in the size of the radio buttons when submitting the question on desktop & tablet, and when the icon appears when in a valid or invalid state.
- For the 5th question of the exit quiz, there is some shift in the size of the checkboxes when submitting the question on desktop & tablet, and when the icon appears when in a valid or invalid state.
- Checking if part of the scope of this ticket - the feedback is missing from the bottom nav when a quiz answer is correct or incorrect. E.g: 
Screenshot 2024-01-23 at 2 39 32 pm
- Can we check which FontTokens the question stem is using? It should use`heading-4` for desktop & tablet, and `heading-6` for mobile.
- Checking if part of the scope of this ticket - will the 'Select one answer' text be a part of another ticket for MCQ questions? E.g: 
Screenshot 2024-01-23 at 1 31 43 pm
- Checking if part of the scope of this ticket - will the counter on the top right be a part of another ticket? E.g: 
Screenshot 2024-01-23 at 1 35 45 pm
- Just to note that the text in the top header and the button text label will be updated as part of the suggestions from Alex that she shared with me on Friday. I'll create a separate ticket for those text updates for the next sprint once they are finalised and reflected in the designs.

That's all the feedback I'll add for now as it is difficult to review further until those updates are made. I'll have another review once updated. Thanks

File: src/components/PupilComponents/QuizRenderer/QuizRenderer.tsx
Comment (benprotheroe) [reply src/components/PupilComponents/QuizRenderer/QuizRenderer.tsx:75]: good idea

File: src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx
Comment (benprotheroe) [reply src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx]: @kimon-satan  youre taking this one right?

**Comment by rightsaidjames** [comment]
Bugs/design differences spotted so far:
* In the feedback area, correct questions should have a `Well done!` feedback text below the `Correct` feedback heading. [Figma link](https://www.figma.com/file/G7z5ufo2T0CHk14CkadeuB/Pupils-high-fidelity-designs?type=design&node-id=3093%3A76801&mode=dev)
* Also, incorrect answers should have the correct answer(s) displayed below the `Incorrect` feedback heading in the format `Correct answer: [correct answer(s)]`. [Figma link](https://www.figma.com/file/G7z5ufo2T0CHk14CkadeuB/Pupils-high-fidelity-designs?type=design&node-id=3038%3A174627&mode=dev)
* Checkbox questions where the user should give feedback of `Almost correct` when the user has selected not enough right answers or a mixture of right and wrong answers, as well as displaying the full list of correct answers in the format `Correct answer: [correct answer(s)]`. [Figma link](https://www.figma.com/file/G7z5ufo2T0CHk14CkadeuB/Pupils-high-fidelity-designs?type=design&node-id=3038%3A178417&mode=dev)
* Also for checkbox questions, when a user selects all of the correct answers **and** one or more incorrect answers, the feedback is incorrectly given as 'Correct' and the user's score increases by 1 on the overview page! 
* Once again on the checkbox questions, the help text says `Select 2 answer` instead of `Select 2 answers` - I'll leave it up to you if you want to do some clever auto-pluralisation logic here, or just make it so that checkbox questions are always `Select X answers` (because a checkbox question with only 1 answer should be a radio button question). [Figma link](https://www.figma.com/file/G7z5ufo2T0CHk14CkadeuB/Pupils-high-fidelity-designs?type=design&node-id=3038%3A175956&mode=dev)

Screenshots:
Screenshot 2024-01-25 at 11 26 03
Screenshot 2024-01-25 at 11 26 37
Screenshot 2024-01-25 at 11 26 45
Screenshot 2024-01-25 at 11 28 48
Screenshot 2024-01-25 at 11 27 36
Screenshot 2024-01-25 at 11 30 10
Screenshot 2024-01-25 at 11 30 27
Screenshot 2024-01-25 at 11 28 18

**Comment by messmike** [comment]
Thanks @benprotheroe. Besides what @rightsaidjames added above, here is the design feedback:

- For short answer questions, the 'Type your answer' text label looks like it is a different styling to the designs. Should be 'heading-light-6' for desktop/tablet and 'heading-light-7' for mobile (should be the same for the 'Select one answer' or the 'Select 2 answers' text for MCQ questions.

- For all starter quiz MCQ questions, the space between the 'Select one answer' and the 'Select 2 answers' text and the radio buttons/checkboxes group needs to decrease to use 'space-between-m' across all breakpoints. This should also be the same for short answer questions with the 'Type your answer' text label. E.g: 
Screenshot 2024-01-25 at 12 43 05 pm
Screenshot 2024-01-25 at 12 43 23 pm

The top nav on mobile should display the question counter when in the starter/exit quiz sections. E.g: 
Screenshot 2024-01-25 at 12 46 02 pm

**Comment by rightsaidjames** [comment]
All of the fixes I raised look good for me, except that the Feedback for partially correct checkbox answers is still 'Incorrect' rather than 'Almost correct':
image

Here's the rest of my screenshots showing the fixes:
image
image
image
image

**Review by rightsaidjames — APPROVED** [review]
Confirmed that 'Almost correct' state is present and correct - all good from me now :) 
Screenshot 2024-01-26 at 09 32 53

**Review by messmike — APPROVED** [review]

File: src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx
Comment (carlmw) [inline src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx:63]: 🏄🏼 fwiw i'd have extracted this into a little function `pickFeedback` so that most of the logic is outside the JSX.

```jsx
<OakLessonBottomNav
  feedback={isFeedbackMode ? pickFeedback(grade) : null}
  ...
</OakLessonBottomNav>
```

```typescript
function pickFeedback(grade: number) {
  switch (grade) {
    case 1:
      return 'correct';
    case 0:
      return 'incorrect';
    case 0.5:
      return 'partially-correct';
    default:
      return null;
  }
}
```

when i'm reading a large block of JSX it's really helpful if I can quickly work out the intent from names of things without having to follow the implementation all the way through.

**Review by carlmw — CHANGES_REQUESTED** [review]
No blockers from me, just a few things i've spotted otherwise looking good. Comments with a 🏄🏼 are just opinions which you can feel free to ignore or counter 😉.

Ping me directly when you want another review and i'll get on on immediately.

File: src/components/PupilComponents/QuizQuestionStem/QuizQuestionStem.test.tsx
Comment (benprotheroe) [reply src/components/PupilComponents/QuizQuestionStem/QuizQuestionStem.test.tsx:39]: have removed i think sometime the formatter adds them

File: src/components/PupilComponents/QuizMCQSingleAnswer/QuizMCQSingleAnswer.tsx
Comment (benprotheroe) [reply src/components/PupilComponents/QuizMCQSingleAnswer/QuizMCQSingleAnswer.tsx:85]: prettier

File: src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx
Comment (benprotheroe) [reply src/components/PupilViews/PupilQuiz/PupilQuiz.view.tsx:74]: switch statement?

File: src/components/PupilComponents/QuizEngineProvider/QuizEngineProvider.tsx
Comment (carlmw) [reply src/components/PupilComponents/QuizEngineProvider/QuizEngineProvider.tsx]: Yeah something like that. Though at that point it could be extracted into a little function at the bottom of the module (with early returns instead of breaks). Means your component doesn't read like a 10 page essay. Helps anyone that comes to this down the road read your intent without having to comprehend all the code. Also keeps diffs minimal when indentation changes!

**Review by carlmw — APPROVED** [review]
This one has been a journey 😉 . :shipit:

**Pull request merged**
